File: java/regular_expression_matching.java
public class Solution {
    boolean isMatch(char[] s, int i, char[] p, int j) {
        if (j==p.length) return i==s.length;
        if (j==p.length-1 || p[j+1]!='*') {
            return (i<s.length && (p[j]=='.' || s[i]==p[j]) && isMatch(s, i+1, p, j+1));
        } else {
            while (i<s.length && (p[j]=='.' || s[i]==p[j])) {
                if (isMatch(s, i, p, j+2)==true) return true;
                i++;
            }
            return isMatch(s, i, p, j+2);
        }
    }
    
    public boolean isMatch(String s, String p) {
        return isMatch(s.toCharArray(), 0, p.toCharArray(), 0);
    }
}

File: golang/regular-expression-matching.go
func isMatchHelper(s string, p string, i int, j int) bool {
	if j == len(p) {
		return i == len(s)
	}

	if j == len(p)-1 || p[j+1] != '*' {
		return i < len(s) && (s[i] == p[j] || p[j] == '.') && isMatchHelper(s, p, i+1, j+1)
	}

	for i < len(s) && (s[i] == p[j] || p[j] == '.') {
		if isMatchHelper(s, p, i, j+2) {
			return true
		}
		i++
	}
	return isMatchHelper(s, p, i, j+2)
}

func isMatch(s string, p string) bool {
	return isMatchHelper(s, p, 0, 0)
}

